R/misc.R

```diff
@@ -475,39 +475,93 @@ load_regional_functional_data <- function(...) {
                               variable_name_obj = "variant_names",
                               twas_weights_table = "twas_weights") {
   ## Internal function to load and validate data from RDS files
-  validate_data <- function(weight_db_file, condition, variable_name_obj) {
-    ## Check if the specified region and condition are available in all files
-    if (!(condition %in% names(weight_db_file[[1]]))) {
-        stop("The specified condition is not available in weight RDS file.")
+  load_and_validate_data <- function(weight_db_files, conditions, variable_name_obj) {
+    all_data <- lapply(weight_db_files, readRDS)
+    unique_regions <- unique(unlist(lapply(all_data,function(data) names(data))))
+    # Check if region from all RDS files are the same
+    if (length(unique_regions) != 1) {
+      stop("The RDS files do not refer to the same region.")
+    } else {
+    # Assuming all data refer to the same region, now combine data by conditions
+      combined_all_data <- do.call("c", lapply(all_data, function(data) data[[1]]))
     }
-    if (!is.null(variable_name_obj) && !(variable_name_obj %in% names(weight_db_file[[1]][[condition]]))) {
-        stop("The specified variable_name_obj is not available in weight RDS file.")
+    # Set default for 'conditions' if they are not specified
+    if (is.null(conditions)) {
+    conditions <- names(combined_all_data)
+    }                                           
+    ## Check if the specified condition and variable_name_obj are available in all files
+    if (!all(conditions %in% names(combined_all_data))) {
+        stop("The specified condition is not available in all RDS files.")
+      }
+    for (data in seq_along(combined_all_data))                                       
+    {
+        if (!is.null(variable_name_obj) && !(variable_name_obj %in% names(combined_all_data[[data]]))) {
+            stop("The specified variable_name_obj is not available in all RDS files.")
+        }
+    }
+    return(combined_all_data)
+  }
+  # Internal function to align and merge weight matrices
+  align_and_merge <- function(weights_list, variable_objs) {
+    # Get the complete list of variant names across all files
+    all_variants <- unique(unlist(variable_objs))
+    consolidated_list = list()
+    # Fill the matrix with weights, aligning by variant names
+    for (i in seq_along(weights_list)) {
+      # Initialize the temp matrix with zeros
+      existing_colnames <- character(0)
+      temp_matrix <- matrix(0, nrow = length(all_variants), ncol = ncol(weights_list[[i]]))
+      rownames(temp_matrix) <- all_variants
+      idx <- match(variable_objs[[i]], all_variants)
+      temp_matrix[idx, ] <- weights_list[[i]]
+      # Ensure no duplicate column names
+      new_colnames <- colnames(weights_list[[i]])
+      if (any(duplicated(c(existing_colnames, new_colnames)))) {
+        stop("Duplicate column names detected during merging process.")
```

Comment:
Please change the error message to reporting the duplicated names. Otherwise this error message would be very difficult to understand by analysts later.